recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift

```diff
@@ -0,0 +1,335 @@
+
+        let modelAccess = selectedModelUrl!.startAccessingSecurityScopedResource()
+        if !modelAccess {
+            errorMessage = "Can't get permissions to access model file"
+            enableLoadModelButton = true
+            return
+        }
+
+        DispatchQueue.global(qos: .userInitiated).async { [self] in
+            let setStatusText = {(_ msg: String) in
+                DispatchQueue.main.async { [self] in
+                    modelLoadStatusText = msg
+                }
+            }
+            do {
+                setStatusText("Loading Porcupine...")
+                porcupine = try Porcupine(accessKey: ACCESS_KEY, keyword: .picovoice)
+
+                setStatusText("Loading Cheetah...")
+                let cheetahModelPath = Bundle(for: type(of: self)).path(forResource: "cheetah_params", ofType: "pv")!
+                cheetah = try Cheetah(accessKey: ACCESS_KEY, modelPath: cheetahModelPath)
+
+                setStatusText("Loading picoLLM...")
+                picollm = try PicoLLM(accessKey: ACCESS_KEY, modelPath: selectedModelUrl!.path)
+                dialog = try picollm!.getDialog()
+
+                setStatusText("Loading Orca...")
+                let orcaModelPath = Bundle(for: type(of: self)).path(forResource: "orca_params_female", ofType: "pv")!
+                orca = try Orca(accessKey: ACCESS_KEY, modelPath: orcaModelPath)
+
+                setStatusText("Loading Audio Player...")
+                audioStream = try AudioPlayerStream(sampleRate: Double(self.orca!.sampleRate!))
+
+                setStatusText("Loading Voice Processor...")
+                VoiceProcessor.instance.addFrameListener(VoiceProcessorFrameListener(audioCallback))
+                VoiceProcessor.instance.addErrorListener(VoiceProcessorErrorListener(errorCallback))
+                try VoiceProcessor.instance.start(
+                    frameLength: Porcupine.frameLength,
+                    sampleRate: Porcupine.sampleRate)
+
+                DispatchQueue.main.async { [self] in
+                    enginesLoaded = true
+                }
+            } catch {
+                DispatchQueue.main.async { [self] in
+                    unloadEngines()
+                    errorMessage = "\(error.localizedDescription)"
+                }
+            }
+
+            DispatchQueue.main.async { [self] in
+                selectedModelUrl!.stopAccessingSecurityScopedResource()
+
+                modelLoadStatusText = ViewModel.modelLoadStatusTextDefault
+                enableLoadModelButton = true
+            }
+        }
+    }
+
+    public func unloadEngines() {
+        do {
+            try VoiceProcessor.instance.stop()
+            VoiceProcessor.instance.clearFrameListeners()
+            VoiceProcessor.instance.clearErrorListeners()
+        } catch {
+            DispatchQueue.main.async { [self] in
+                errorMessage = "\(error.localizedDescription)"
+            }
+        }
```

Comment:
Does the voice processor record even when we're generating the completion and doing TTS? If that's the case, I think it would be better to start it on wakeword detect and stop it after cheetah finishes. This way we're running fewer threads and closing the mic stream when we're not using it

Reply:
Actually, on iOS it seems that stopping and starting voice processor is a fairly expensive call. It seems to interrupt audio playback which makes a poor user experience.

Reply:
That's odd... it shouldn't be that expensive and it really shouldn't affect the audio playback since stopping the input stream should happen entirely before the output stream starts. 

State should be:
`recording start -> wake word -> stt -> recording stop -> llm generate -> tts -> ...back to beginning`

It kind of sends a weird message to devs looking at this if we keep the mic stream open when it is not using audio. Especially if they use it as a basis for a project where they introduce even more app logic to the llm/tts portion. If there is a way to keep it open only when we need it, that would be preferable.